vector/src/main/java/im/vector/app/features/home/room/detail/composer/MessageComposerFragment.kt

```diff
@@ -0,0 +1,803 @@
+
+        attachmentsHelper = AttachmentsHelper(requireContext(), this, buildMeta).register()
+
+        setupComposer()
+        setupEmojiButton()
+
+        messageComposerViewModel.observeViewEvents {
+            when (it) {
+                is MessageComposerViewEvents.JoinRoomCommandSuccess -> handleJoinedToAnotherRoom(it)
+                is MessageComposerViewEvents.SlashCommandConfirmationRequest -> handleSlashCommandConfirmationRequest(it)
+                is MessageComposerViewEvents.SendMessageResult -> renderSendMessageResult(it)
+                is MessageComposerViewEvents.ShowMessage -> showSnackWithMessage(it.message)
+                is MessageComposerViewEvents.ShowRoomUpgradeDialog -> handleShowRoomUpgradeDialog(it)
+                is MessageComposerViewEvents.AnimateSendButtonVisibility -> handleSendButtonVisibilityChanged(it)
+                is MessageComposerViewEvents.OpenRoomMemberProfile -> openRoomMemberProfile(it.userId)
+                is MessageComposerViewEvents.VoicePlaybackOrRecordingFailure -> {
+                    if (it.throwable is VoiceFailure.UnableToRecord) {
+                        onCannotRecord()
+                    }
+                    showErrorInSnackbar(it.throwable)
+                }
+                is MessageComposerViewEvents.InsertUserDisplayName -> insertUserDisplayNameInTextEditor(it.userId)
+            }
+        }
+
+        messageComposerViewModel.onEach(MessageComposerViewState::sendMode, MessageComposerViewState::canSendMessage) { mode, canSend ->
+            if (!canSend.boolean()) {
+                return@onEach
+            }
+            when (mode) {
+                is SendMode.Regular -> renderRegularMode(mode.text.toString())
+                is SendMode.Edit -> renderSpecialMode(mode.timelineEvent, R.drawable.ic_edit, R.string.edit, mode.text.toString())
+                is SendMode.Quote -> renderSpecialMode(mode.timelineEvent, R.drawable.ic_quote, R.string.action_quote, mode.text.toString())
+                is SendMode.Reply -> renderSpecialMode(mode.timelineEvent, R.drawable.ic_reply, R.string.reply, mode.text.toString())
+                is SendMode.Voice -> renderVoiceMessageMode(mode.text)
+            }
+        }
+
+        if (savedInstanceState != null) {
+            handleShareData()
+        }
+    }
+
+    override fun onPause() {
+        super.onPause()
+
+        if (withState(messageComposerViewModel) { it.isVoiceRecording } && requireActivity().isChangingConfigurations) {
+            // we're rotating, maintain any active recordings
+        } else {
+            messageComposerViewModel.handle(MessageComposerAction.OnEntersBackground(views.composerLayout.text.toString()))
+        }
+    }
+
+    override fun onDestroyView() {
+        super.onDestroyView()
+
+        autoCompleter.clear()
+        messageComposerViewModel.endAllVoiceActions()
```

Comment:
Maybe move this line to `VoiceRecorderFragment.onDestroyView()` ?

Reply:
Actually, I think that line is not needed. Leaving the screen by either going back, leaving the app or going to a new screen sends the `MessageComposerAction.OnEntersBackground` action, that always stops either the playback or the recording of a voice message.

Reply:
Also, I think I left if there because it would also stop any sent/received voice messages that were being played from the timeline items.

Reply:
OK, thanks for checking.